### ISSUE TITLE
Test failure t/re_fold_grind.t with bleadperl

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#11258]
Migrated from [rt.perl.org#88744](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=88744) (status was 'resolved')

Searchable as RT88744$

**Comment by p5pRT** [comment]
### From ambrus@math.bme.hu

#### Created by ambrus@math.bme.hu




I've got a failure in the core test t/re/fold\_grind\.t when building blead\-perl
on amd\-64\.

The build is today's perl blead \(described "v5\.13\.11\-455\-g40f026236b99"\)
with no special configuration \(you can see the configuration at the bottom
of this mail\) on an amd64\-linux system that's based on debian squeeze but
has lots of new software such as gcc 4\.5\.1\.  The build seems to mostly work&#8203;:
this test file is the only one that has failed\.

In case anyone cares\, I've also got the same test failure from building perl
5\.13\.11 development release on the same machine; but building 5\.13\.9 had no
test failures\.

The whole output of the test is too long to send here\, but here's some
selected lines\.


&nbsp;      1 \# Failed at re/fold\_grind\.t line 70
&nbsp;      5 \# Failed at re/fold\_grind\.t line 579
&nbsp;    598 \# Failed at re/fold\_grind\.t line 579
&nbsp;    599 \# Failed at re/fold\_grind\.t line 596
&nbsp;    600 \#      got "126"
&nbsp;    601 \# expected "720"
&nbsp;  10749 ok 1132 \- Subtests okay for charset=aa\, utf8\_pattern=1
&nbsp;  10750 not ok 1133 \- my $c = "\\x\{00C0\}\\x\{00E0\}"; my $p =
qr/\(?l&#8203;:^\(\\x\{00E0\}\)\\1$\)/i; $c \!~ $p;
&nbsp;  10751 not ok 1134 \- my $c = "\\x\{00C0\}\\x\{00E0\}"; my $p =
qr/\(?l&#8203;:^\(?\<grind>\\x\{00E0\}\)\\k\<grind>$\)/i; $c \!~ $p;
&nbsp;  10752 not ok 1135 \- my $c = "\\x\{00E0\}\\x\{00C0\}"; my $p =
qr/\(?l&#8203;:^\(\\x\{00E0\}\)\\1$\)/i; $c \!~ $p;
&nbsp;  10753 not ok 1136 \- my $c = "\\x\{00E0\}\\x\{00C0\}"; my $p =
qr/\(?l&#8203;:^\(?\<grind>\\x\{00E0\}\)\\k\<grind>$\)/i; $c \!~ $p;
&nbsp;  10754 not ok 1137 \- my $c = "\\x\{00C0\}"; my $p = qr/\(?l&#8203;:\\x\{00E0\}\)/i; $c \!~ $p;
&nbsp;  10755 not ok 1138 \- my $c = "\\x\{00C0\}"; my $p = qr/\(?l&#8203;:\\x\{00E0\}$\)/i;
$c \!~ $p;
&nbsp;  10756 not ok 1139 \- my $c = "\\x\{00C0\}"; my $p = qr/\(?l&#8203;:^\\x\{00E0\}\)/i;
$c \!~ $p;
&nbsp;  10757 not ok 1140 \- my $c = "\\x\{00C0\}"; my $p =
qr/\(?l&#8203;:^\\x\{00E0\}$\)/i; $c \!~ $p;
&nbsp;  11346 not ok 1729 \- my $c = "\_\\x\{00C0\}\_"; my $p =
qr/\(?l&#8203;:^\_\(\(\[^\\x\{00E0\}\_\]\)\+\,?\)\{1\,3\}\_\)/i; $c =~ $p;
&nbsp;  11347 not ok 1730 \- my $c = "\_\\x\{00C0\}\_"; my $p =
qr/\(?l&#8203;:^\_\(\(\[^\\x\{00E0\}\_\]\)\+\,?\)\{1\,3\}\_$\)/i; $c =~ $p;
&nbsp;  11348 not ok 1731 \- Subtests okay for charset=l\, utf8\_pattern=0
&nbsp;  11349 not ok 1732 \- my $c = "\\x\{00C0\}\\x\{00E0\}"; my $p =
qr/\(?l&#8203;:^\(\\x\{00E0\}\)\\1$\)/i; utf8&#8203;::upgrade\($p\); $c \!~ $p;
&nbsp;  11350 not ok 1733 \- my $c = "\\x\{00C0\}\\x\{00E0\}"; my $p =
qr/\(?l&#8203;:^\(?\<grind>\\x\{00E0\}\)\\k\<grind>$\)/i; utf8&#8203;::upgrade\($p\); $c \!~ $p;
&nbsp;  11351 not ok 1734 \- my $c = "\\x\{00E0\}\\x\{00C0\}"; my $p =
qr/\(?l&#8203;:^\(\\x\{00E0\}\)\\1$\)/i; utf8&#8203;::upgrade\($p\); $c \!~ $p;
&nbsp;  11352 not ok 1735 \- my $c = "\\x\{00E0\}\\x\{00C0\}"; my $p =
qr/\(?l&#8203;:^\(?\<grind>\\x\{00E0\}\)\\k\<grind>$\)/i; utf8&#8203;::upgrade\($p\); $c \!~ $p;
&nbsp;  11353 not ok 1736 \- my $c = "\\x\{00C0\}"; my $p = qr/\(?l&#8203;:\\x\{00E0\}\)/i;
utf8&#8203;::upgrade\($p\); $c \!~ $p;
&nbsp;  11354 not ok 1737 \- my $c = "\\x\{00C0\}"; my $p = qr/\(?l&#8203;:\\x\{00E0\}$\)/i;
utf8&#8203;::upgrade\($p\); $c \!~ $p;
&nbsp;  11355 not ok 1738 \- my $c = "\\x\{00C0\}"; my $p = qr/\(?l&#8203;:^\\x\{00E0\}\)/i;
utf8&#8203;::upgrade\($p\); $c \!~ $p;
&nbsp;  11356 not ok 1739 \- my $c = "\\x\{00C0\}"; my $p =
qr/\(?l&#8203;:^\\x\{00E0\}$\)/i; utf8&#8203;::upgrade\($p\); $c \!~ $p;
&nbsp;  23048 ok 13431 \- my $c = "\\x\{0053\}\\x\{0054\}\\x\{0073\}\\x\{0074\}"; my $p =
qr/\(?u&#8203;:^\(?\<grind>\\x\{0053\}\\x\{0054\}\)\\k\<grind>$\)/i; utf8&#8203;::upgrade\($c\);
utf8&#8203;::upgrade\($p\); $c =~ $p;
&nbsp;  23049 1\.\.13431
&nbsp;  23050 Failed 9584/13431 subtests
&nbsp;  23051
&nbsp;  23052 Test Summary Report
&nbsp;  23053 \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
&nbsp;  23054 re/fold\_grind\.t \(Wstat&#8203;: 0 Tests&#8203;: 13431 Failed&#8203;: 9584\)
&nbsp;  23055   Failed tests&#8203;:  1133\-3528\, 3609\-6004\, 6065\-8460\, 8561\-10956
&nbsp;  23056 Files=1\, Tests=13431\, 16 wallclock secs \( 1\.21 usr  0\.01 sys \+
16\.24 cusr  0\.17 csys = 17\.63 CPU\)
&nbsp;  23057 Result&#8203;: FAIL


Thanks\,
Ambrus

Ps\. for those who's skimmed to the bottom\, this is not from a stable
release\, but from perl\-blead some weeks before the release\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

Site configuration information for perl 5.14.0:

Configured by ambrus at Sun Apr 17 10:43:00 CEST 2011.

Summary of my perl5 (revision 5 version 14 subversion 0) configuration:
  Commit id: 40f026236b9959b7ad3260fedc6c66cd30bb7abc
  Platform:
    osname=linux, osvers=2.6.37, archname=x86_64-linux
    uname='linux king 2.6.37 #6 smp sun mar 13 20:15:05 cet 2011
x86_64 gnulinux '
    config_args='-Dprefix=~/local/perlblead -Dusedevel -d'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=undef, usemultiplicity=undef
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=define, use64bitall=define, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-fno-strict-aliasing -pipe -fstack-protector
-I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2',
    cppflags='-fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include'
    ccversion='', gccversion='4.5.1', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t',
lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -fstack-protector -L/usr/local/lib'
    libpth=/usr/local/lib /lib/../lib64/ /usr/lib/../lib64/ /lib/
/usr/lib/ /lib /usr/lib /lib64 /usr/lib64 /usr/local/lib64
    libs=-lnsl -ldl -lm -lcrypt -lutil -lc
    perllibs=-lnsl -ldl -lm -lcrypt -lutil -lc
    libc=/lib/libc-2.11.2.so, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.11.2'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -O2 -L/usr/local/lib
-fstack-protector'

Locally applied patches:
    RC0


@INC for perl 5.14.0:
    /home/ambrus/local/perlblead/lib/site_perl/5.14.0/x86_64-linux
    /home/ambrus/local/perlblead/lib/site_perl/5.14.0
    /home/ambrus/local/perlblead/lib/5.14.0/x86_64-linux
    /home/ambrus/local/perlblead/lib/5.14.0
    .


Environment for perl 5.14.0:
    HOME=/home/ambrus
    LANG (unset)
    LANGUAGE (unset)
    LC_CTYPE=hu_HU
    LD_LIBRARY_PATH=/home/ambrus/local/lib/
    LOGDIR (unset)
    PATH=/home/ambrus/local/bin:/usr/local/bin:/usr/bin:/bin:/usr/bin/X11:/usr/games
    PERL_BADLANG (unset)
    SHELL=/usr/local/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @obra




On Sun 17\.Apr'11 at  2&#8203;:54&#8203;:51 \-0700\, Zsban Ambrus wrote&#8203;:
> 
> The build is today's perl blead \(described "v5\.13\.11\-455\-g40f026236b99"\)
> with no special configuration \(you can see the configuration at the bottom
> of this mail\) on an amd64\-linux system that's based on debian squeeze but
> has lots of new software such as gcc 4\.5\.1\.  The build seems to mostly work&#8203;:
> this test file is the only one that has failed\.
> 
> In case anyone cares\, I've also got the same test failure from building perl
> 5\.13\.11 development release on the same machine; but building 5\.13\.9 had no
> test failures\.

Are you up for running a bisect? This \_doesn't\_ fail for me with gcc
4\.5\.2 on Ubuntu 11\.04 on x86\_64\.

Best\,
Jesse

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @khwilliamson

On 04/17/2011 08&#8203;:44 PM\, Jesse Vincent wrote&#8203;:
>
>
>
> On Sun 17\.Apr'11 at  2&#8203;:54&#8203;:51 \-0700\, Zsban Ambrus wrote&#8203;:
>>
>> The build is today's perl blead \(described "v5\.13\.11\-455\-g40f026236b99"\)
>> with no special configuration \(you can see the configuration at the bottom
>> of this mail\) on an amd64\-linux system that's based on debian squeeze but
>> has lots of new software such as gcc 4\.5\.1\.  The build seems to mostly work&#8203;:
>> this test file is the only one that has failed\.
>>
>> In case anyone cares\, I've also got the same test failure from building perl
>> 5\.13\.11 development release on the same machine; but building 5\.13\.9 had no
>> test failures\.
>
> Are you up for running a bisect? This \_doesn't\_ fail for me with gcc
> 4\.5\.2 on Ubuntu 11\.04 on x86\_64\.
>
> Best\,
> Jesse
>


All the tests that were in the sample have the /l modifier; my guess is 
that all the failures do too; perhaps Ambrus can verify that\.  Further 
all the failures that were sent to us look like \\xe0 and \\xc0 are 
considered folds of each other\, which they aren't supposed to be in the 
C locale\.

If so\, this looks to be a bug on the target system\-\- a faultily 
implemented C locale\, which we have seen examples of already; I think it 
was netbsd and vms\.  And I created sort of a work\-around\, but I didn't 
do so for this \.t as the previous problems didn't have to do with folding\.

I can envision that someone has changed the C locale on a Hungarian 
machine to not do the standard thing\, in order to make the local code 
work better when people forget to do the proper setlocale\.

An experiment that Ambrus could do would be to manually run 
t/re/charset\.t and see if there are any tests of /l\.  If not\, it means 
that the \.t found the locale to be buggy\.  Something like that could be 
done here\.  The problem with it is we don't know for sure if the locale 
is bad or our code is\.

**Comment by p5pRT** [comment]
### From ambrus@math.bme.hu

On Sun\, Apr 17\, 2011 at 11&#8203;:54 AM\, Zsbán Ambrus \<ambrus@&#8203;math\.bme\.hu> wrote&#8203;:
> I've got a failure in the core test t/re/fold\_grind\.t when building blead\-perl
> on amd\-64\.

I found a lead&#8203;: the test passes if I unset all locale environment
variables when running the test\.  I got the error with my usual locale
settings for an iso\-8859\-2 terminal&#8203;: LC\_CTYPE=hu\_HU set \(this is a
locale with character set iso\-8859\-2; all other locale variables are
unset\)\.

On Mon\, Apr 18\, 2011 at 7&#8203;:32 AM\, Karl Williamson
\<public@&#8203;khwilliamson\.com> wrote&#8203;:
> All the tests that were in the sample have the /l modifier; my guess is that
> all the failures do too; perhaps Ambrus can verify that\.

Yes\, all the failures have the l modifier\.

Ambrus

**Comment by p5pRT** [comment]
### From ambrus@math.bme.hu

> On Sun\, Apr 17\, 2011 at 11&#8203;:54 AM\, Zsbán Ambrus \<ambrus@&#8203;math\.bme\.hu> wrote&#8203;:
> I've got a failure in the core test t/re/fold\_grind\.t when building blead\-perl
> on amd\-64\.

I found a lead&#8203;: the test passes if I unset all locale environment
variables when running the test\.  I got the error with my usual locale
settings for an iso\-8859\-2 terminal&#8203;: LC\_CTYPE=hu\_HU set \(this is a
locale with character set iso\-8859\-2; all other locale variables are
unset\)\.

On Mon\, Apr 18\, 2011 at 7&#8203;:32 AM\, Karl Williamson
\<public@&#8203;khwilliamson\.com> wrote&#8203;:
> All the tests that were in the sample have the /l modifier; my guess is that
> all the failures do too; perhaps Ambrus can verify that\.

Yes\, all the failures have the l modifier\.

Ambrus

**Comment by p5pRT** [comment]
### From ambrus@math.bme.hu

Sorry\, this was meant to be a reply to \[perl \#88744\] instead of a new
report\, I messed up\.

Ambrus

**Comment by p5pRT** [comment]
### From @khwilliamson

On 04/18/2011 02&#8203;:09 AM\, Zsban Ambrus \(via RT\) wrote&#8203;:
> \# New Ticket Created by  Zsban Ambrus
> \# Please include the string&#8203;:  \[perl \#88822\]
> \# in the subject line of all future correspondence about this issue\.
> \#\<URL&#8203;: http&#8203;://rt\.perl\.org/rt3/Ticket/Display\.html?id=88822>
>
>
> On Sun\, Apr 17\, 2011 at 11&#8203;:54 AM\, Zsbán Ambrus\<ambrus@&#8203;math\.bme\.hu>  wrote&#8203;:
>> I've got a failure in the core test t/re/fold\_grind\.t when building blead\-perl
>> on amd\-64\.
>
> I found a lead&#8203;: the test passes if I unset all locale environment
> variables when running the test\.  I got the error with my usual locale
> settings for an iso\-8859\-2 terminal&#8203;: LC\_CTYPE=hu\_HU set \(this is a
> locale with character set iso\-8859\-2; all other locale variables are
> unset\)\.
>
> On Mon\, Apr 18\, 2011 at 7&#8203;:32 AM\, Karl Williamson
> \<public@&#8203;khwilliamson\.com>  wrote&#8203;:
>> All the tests that were in the sample have the /l modifier; my guess is that
>> all the failures do too; perhaps Ambrus can verify that\.
>
> Yes\, all the failures have the l modifier\.
>
> Ambrus
>
>

Could you try out the attached patch\, and report the results?

**Comment by p5pRT** [comment]
### From @khwilliamson

<details><summary>0003-PATCH-perl-88822-Re-Test-failure-t-re_fold_grind.t-w.patch</summary>

```diff
From 1f101cc01524e9d86484bbb0bda406d7072a3927 Mon Sep 17 00:00:00 2001
From: Karl Williamson <public@khwilliamson.com>
Date: Mon, 18 Apr 2011 12:16:29 -0600
Subject: [PATCH 3/3] PATCH: [perl #88822] Re: Test failure t/re_fold_grind.t with bleadperl

---
 t/re/fold_grind.t |   13 ++++++++++++-
 1 files changed, 12 insertions(+), 1 deletions(-)

diff --git a/t/re/fold_grind.t b/t/re/fold_grind.t
index 82ca6ad..8e79ed1 100644
--- a/t/re/fold_grind.t
+++ b/t/re/fold_grind.t
@@ -223,7 +223,18 @@ sub pairs (@) {
 
 my @charsets = qw(d u a aa);
 my $current_locale = POSIX::setlocale( &POSIX::LC_ALL, "C") // "";
-push @charsets, 'l' if $current_locale eq 'C';
+if ($current_locale eq 'C') {
+    use locale;
+
+    # Some locale implementations don't have the 128-255 characters all
+    # mean nothing.  Skip the locale tests in that situation.
+    for my $i (128 .. 255) {
+        my $char = chr($i);
+        goto bad_locale if uc($char) ne $char || lc($char) ne $char;
+    }
+    push @charsets, 'l';
+bad_locale:
+}
 
 # Finally ready to do the tests
 my $count=0;
-- 
1.7.1


```
</details>

**Comment by p5pRT** [comment]
### From ambrus@math.bme.hu

On Mon\, Apr 18\, 2011 at 8&#8203;:19 PM\, Karl Williamson
\<public@&#8203;khwilliamson\.com> wrote&#8203;:
> On 04/18/2011 02&#8203;:09 AM\, Zsban Ambrus \(via RT\) wrote&#8203;:
>> On Sun\, Apr 17\, 2011 at 11&#8203;:54 AM\, Zsbán Ambrus\<ambrus@&#8203;math\.bme\.hu>  wrote&#8203;:
>>>
>>> I've got a failure in the core test t/re/fold\_grind\.t when building
>>> blead\-perl
>>> on amd\-64\.
>>
>> I found a lead&#8203;: the test passes if I unset all locale environment
>> variables when running the test\.  I got the error with my usual locale
>> settings for an iso\-8859\-2 terminal&#8203;: LC\_CTYPE=hu\_HU set \(this is a
>> locale with character set iso\-8859\-2; all other locale variables are
>> unset\)\.
>>
>> On Mon\, Apr 18\, 2011 at 7&#8203;:32 AM\, Karl Williamson
>> \<public@&#8203;khwilliamson\.com>  wrote&#8203;:
>>>
>>> All the tests that were in the sample have the /l modifier; my guess is
>>> that
>>> all the failures do too; perhaps Ambrus can verify that\.
>>
>> Yes\, all the failures have the l modifier\.
>
> Could you try out the attached patch\, and report the results?

Applying your patch to the same pre\-RC1 blead makes the test failure go away\.

Further investigations show that that patch makes the locale tests
skipped with the locale environment variable settings I used
\(LC\_CTYPE=hu\_HU; all other unset\)\.  With all locale environment
variables unset\, the locale tests are still ran and everything is
still okay\.  From this\, I've found something alarming&#8203;:

\[am\]king ~/build/perl/perl\-git$ perl \-wE 'use POSIX \(\); BEGIN \{ say
POSIX&#8203;::setlocale\(POSIX&#8203;::LC\_CTYPE\(\)\, "C"\); say
POSIX&#8203;::setlocale\(POSIX&#8203;::LC\_ALL\(\)\, "C"\); \} use locale; say
ord\(lc\(chr\(201\)\)\);'
C
C
201
\[am\]king ~/build/perl/perl\-git$ perl \-wE 'use POSIX \(\); BEGIN \{ say
POSIX&#8203;::setlocale\(POSIX&#8203;::LC\_ALL\(\)\, "C"\); \} use locale; say
ord\(lc\(chr\(201\)\)\);'
C
233
\[am\]king ~/build/perl/perl\-git$

I believe the latter of these is a bug\, and is present not only in
blead but also in stable perl 5\.12\.3\.

I'll have to check whether this is a bug in libc\.

Ambrus

**Comment by p5pRT** [comment]
### From ambrus@math.bme.hu

On Mon\, Apr 25\, 2011 at 6&#8203;:52 PM\, Zsbán Ambrus \<ambrus@&#8203;math\.bme\.hu> wrote&#8203;:
> I'll have to check whether this is a bug in libc\.

I couldn't reproduce a bug with a standalone C program \(without perl\)\.

However\, debugging a newly built non\-optimized perl I found something
strange&#8203;: the return value of the setlocale C function seems wrong at
one point\.


\[am\]king ~/build/perl/perl\-5\.14\.0\-RC1$ perl \-V
Summary of my perl5 \(revision 5 version 12 subversion 3\) configuration&#8203;:

&nbsp;  Platform&#8203;:
&nbsp;    osname=linux\, osvers=2\.6\.34\.1\, archname=x86\_64\-linux
&nbsp;    uname='linux king 2\.6\.34\.1 \#1 smp sat jul 10 18&#8203;:21&#8203;:56 cest 2010
x86\_64 gnulinux '
&nbsp;    config\_args='\-Dinc\_version\_list=5\.12\.2/x86\_64\-linux 5\.12\.2
5\.12\.1/x86\_64\-linux 5\.12\.1 5\.12\.0/x86\_64\-linux 5\.12\.0 \-d'
&nbsp;    hint=recommended\, useposix=true\, d\_sigaction=define
&nbsp;    useithreads=undef\, usemultiplicity=undef
&nbsp;    useperlio=define\, d\_sfio=undef\, uselargefiles=define\, usesocks=undef
&nbsp;    use64bitint=define\, use64bitall=define\, uselongdouble=undef
&nbsp;    usemymalloc=n\, bincompat5005=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='cc'\, ccflags ='\-fno\-strict\-aliasing \-pipe \-fstack\-protector
\-I/usr/local/include \-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64'\,
&nbsp;    optimize='\-O2'\,
&nbsp;    cppflags='\-fno\-strict\-aliasing \-pipe \-fstack\-protector \-I/usr/local/include'
&nbsp;    ccversion=''\, gccversion='4\.5\.1'\, gccosandvers=''
&nbsp;    intsize=4\, longsize=8\, ptrsize=8\, doublesize=8\, byteorder=12345678
&nbsp;    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=16
&nbsp;    ivtype='long'\, ivsize=8\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\,
lseeksize=8
&nbsp;    alignbytes=8\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='cc'\, ldflags =' \-fstack\-protector \-L/usr/local/lib'
&nbsp;    libpth=/usr/local/lib /lib /usr/lib /lib64 /usr/lib64 /usr/local/lib64
&nbsp;    libs=\-lnsl \-ldl \-lm \-lcrypt \-lutil \-lc
&nbsp;    perllibs=\-lnsl \-ldl \-lm \-lcrypt \-lutil \-lc
&nbsp;    libc=/lib/libc\-2\.7\.so\, so=so\, useshrplib=false\, libperl=libperl\.a
&nbsp;    gnulibc\_version='2\.7'
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags='\-Wl\,\-E'
&nbsp;    cccdlflags='\-fPIC'\, lddlflags='\-shared \-O2 \-L/usr/local/lib
\-fstack\-protector'


Characteristics of this binary \(from libperl\)&#8203;:
&nbsp;  Compile\-time options&#8203;: PERL\_DONT\_CREATE\_GVSV PERL\_MALLOC\_WRAP USE\_64\_BIT\_ALL
&nbsp;                        USE\_64\_BIT\_INT USE\_LARGE\_FILES USE\_PERLIO
&nbsp;                        USE\_PERL\_ATOF
&nbsp;  Built under linux
&nbsp;  Compiled at Jan 25 2011 14&#8203;:36&#8203;:36
&nbsp;  @&#8203;INC&#8203;:
&nbsp;    /usr/local/lib/perl5/site\_perl/5\.12\.3/x86\_64\-linux
&nbsp;    /usr/local/lib/perl5/site\_perl/5\.12\.3
&nbsp;    /usr/local/lib/perl5/5\.12\.3/x86\_64\-linux
&nbsp;    /usr/local/lib/perl5/5\.12\.3
&nbsp;    /usr/local/lib/perl5/site\_perl/5\.12\.2/x86\_64\-linux
&nbsp;    /usr/local/lib/perl5/site\_perl/5\.12\.2
&nbsp;    /usr/local/lib/perl5/site\_perl/5\.12\.1/x86\_64\-linux
&nbsp;    /usr/local/lib/perl5/site\_perl/5\.12\.1
&nbsp;    /usr/local/lib/perl5/site\_perl/5\.12\.0/x86\_64\-linux
&nbsp;    /usr/local/lib/perl5/site\_perl/5\.12\.0
&nbsp;    /usr/local/lib/perl5/site\_perl
&nbsp;    \.
\[am\]king ~/build/perl/perl\-5\.14\.0\-RC1$ locale
LANG=
LC\_CTYPE=hu\_HU
LC\_NUMERIC="POSIX"
LC\_TIME="POSIX"
LC\_COLLATE="POSIX"
LC\_MONETARY="POSIX"
LC\_MESSAGES="POSIX"
LC\_PAPER="POSIX"
LC\_NAME="POSIX"
LC\_ADDRESS="POSIX"
LC\_TELEPHONE="POSIX"
LC\_MEASUREMENT="POSIX"
LC\_IDENTIFICATION="POSIX"
LC\_ALL=
\[am\]king ~/build/perl/perl\-5\.14\.0\-RC1$ \./perl \-wE 'use POSIX; use
locale; say setlocale\(LC\_ALL\, "C"\); say ord\(lc\(chr\(201\)\)\);'
C
233
\[am\]king ~/build/perl/perl\-5\.14\.0\-RC1$ \# ^^^ WRONG\, should be 201
\[am\]king ~/build/perl/perl\-5\.14\.0\-RC1$ gdb \./perl
GNU gdb \(GDB\) 7\.1
Copyright \(C\) 2010 Free Software Foundation\, Inc\.
License GPLv3\+&#8203;: GNU GPL version 3 or later \<http&#8203;://gnu\.org/licenses/gpl\.html>
This is free software&#8203;: you are free to change and redistribute it\.
There is NO WARRANTY\, to the extent permitted by law\.  Type "show copying"
and "show warranty" for details\.
This GDB was configured as "x86\_64\-unknown\-linux\-gnu"\.
For bug reporting instructions\, please see&#8203;:
\<http&#8203;://www\.gnu\.org/software/gdb/bugs/>\.\.\.
Reading symbols from /home/ambrus/build/perl/perl\-5\.14\.0\-RC1/perl\.\.\.done\.
\(gdb\) b XS\_POSIX\_setlocale
Function "XS\_POSIX\_setlocale" not defined\.
Make breakpoint pending on future shared library load? \(y or \[n\]\) y

Breakpoint 1 \(XS\_POSIX\_setlocale\) pending\.
\(gdb\) r \-wE 'use POSIX; use locale; say setlocale\(LC\_ALL\, "C"\); say
ord\(lc\(chr\(201\)\)\);'
Starting program&#8203;: /home/ambrus/build/perl/perl\-5\.14\.0\-RC1/perl \-wE
'use POSIX; use locale; say setlocale\(LC\_ALL\, "C"\); say
ord\(lc\(chr\(201\)\)\);'

Breakpoint 1\, XS\_POSIX\_setlocale \(cv=0x894190\) at POSIX\.c&#8203;:2232
2232        dVAR; dXSARGS;
\(gdb\) n
2236        if \(items \< 1 || items > 2\)
\(gdb\)
2239        int    category = \(int\)SvIV\(ST\(0\)\);
\(gdb\)
2245        dXSTARG;
\(gdb\)
2247        if \(items \< 2\)
\(gdb\)
2250            locale = \(char \*\)SvPV\_nolen\(ST\(1\)\);
\(gdb\)
1226        retval = setlocale\(category\, locale\);
\(gdb\)
1227        if \(retval\) \{
\(gdb\) p category
$1 = 6
\(gdb\) p locale
$2 = 0x90a620 "C"
\(gdb\) p retval
$3 = 0x7ffff70cc0f2 "C"
\(gdb\) \# btw\, 6 = LC\_ALL
\(gdb\) n
1230            RETVAL = savepv\(retval\);
\(gdb\)
1232            if \(category == LC\_CTYPE
\(gdb\)
1234            || category == LC\_ALL
\(gdb\)
1240            if \(category == LC\_ALL\)
\(gdb\)
1241                newctype = setlocale\(LC\_CTYPE\, NULL\);
\(gdb\)
1245            new\_ctype\(newctype\);
\(gdb\) p newctype
$4 = 0x7c2780 "hu\_HU"
\(gdb\) \# ^^^ WHY? this should give "C"
\(gdb\) q
A debugging session is active\.

&nbsp;    Inferior 1 \[process 24194\] will be killed\.

Quit anyway? \(y or n\) y
\[am\]king ~/build/perl/perl\-5\.14\.0\-RC1$


Meanwhile\, allow me a question that might not be related but might
clear up something I don't understand\.  In the function
Perl\_upg\_version in file util\.c\, at one point it says this&#8203;:

&nbsp;        char \*loc = setlocale\(LC\_NUMERIC\, "C"\);
&nbsp;        \.\.\.
&nbsp;        setlocale\(LC\_NUMERIC\, loc\);

What's the point of the second setlocale here?  Doesn't it always just
set the C locale again?  Afterall\, the first call to setlocale should
always return "C" \(or something equivalent\) because that's the locale
it sets\.  The setlocale C function doesn't return the old locale\, it
returns the new locale\.  So why is that second call to setlocale
there?  Same goes to function S\_force\_version in file toke\.c\.

Ambrus

**Comment by p5pRT** [comment]
### From @cpansprout

On Mon Apr 18 11&#8203;:20&#8203;:54 2011\, public@&#8203;khwilliamson\.com wrote&#8203;:
> On 04/18/2011 02&#8203;:09 AM\, Zsban Ambrus \(via RT\) wrote&#8203;:
> > \# New Ticket Created by  Zsban Ambrus
> > \# Please include the string&#8203;:  \[perl \#88822\]
> > \# in the subject line of all future correspondence about this issue\.
> > \#\<URL&#8203;: http&#8203;://rt\.perl\.org/rt3/Ticket/Display\.html?id=88822>
> >
> >
> > On Sun\, Apr 17\, 2011 at 11&#8203;:54 AM\, Zsb�n Ambrus\<ambrus@&#8203;math\.bme\.hu>
> wrote&#8203;:
> >> I've got a failure in the core test t/re/fold\_grind\.t when building
> blead\-perl
> >> on amd\-64\.
> >
> > I found a lead&#8203;: the test passes if I unset all locale environment
> > variables when running the test\.  I got the error with my usual
> locale
> > settings for an iso\-8859\-2 terminal&#8203;: LC\_CTYPE=hu\_HU set \(this is a
> > locale with character set iso\-8859\-2; all other locale variables are
> > unset\)\.
> >
> > On Mon\, Apr 18\, 2011 at 7&#8203;:32 AM\, Karl Williamson
> > \<public@&#8203;khwilliamson\.com>  wrote&#8203;:
> >> All the tests that were in the sample have the /l modifier; my
> guess is that
> >> all the failures do too; perhaps Ambrus can verify that\.
> >
> > Yes\, all the failures have the l modifier\.
> >
> > Ambrus
> >
> >
> 
> Could you try out the attached patch\, and report the results?

That has now been applied as 24282c4\.

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From ambrus@math.bme.hu

Bram has found the root cause of the locale problem\.  It turns out it's
a bug in glibc that has been present for a decade now\.  The report and
patch about this is
"http&#8203;://sourceware\.org/bugzilla/show\_bug\.cgi?id=12788"\.  

P5p will need to decide whether the perl core should try to work around
this glibc bug\.
